### ISSUE TITLE
How to configure reply by email

### ISSUE THREAD

**Issue by tomdereub** [YunoHost-Apps/discourse_ynh#2]
Hi,
I don't understand how to configure reply by email with yunohost. This tuto : https://meta.discourse.org/t/set-up-reply-via-email-support/14003 uses a gmail account for that, but there's already a discourse account, isn't it ? Or do I need to create a new yunohost user just for that ?

**Comment by JimboJoe** [comment]
A `discourse` system user is created, but no YunoHost user, hence no e-mail account. I must admit I haven't tested myself yet, but I guess creating a YunoHost dedicated user (to create a mailbox) and following the tutorial to setup the e-mail access should do the trick.
Don't forget to report back, or even propose a Pull Request on the README file if you got it working :wink:

**Comment by tomdereub** [comment]
You say there is no e-mail account, but discourse is able to send mail, so how does it work ?
I'm trying to configure it after creating a dedicated yunohost user, but I don't find if there is a POP service in yunohost ? The documentation just speaks about imap.

**Comment by JimboJoe** [comment]
The Discourse app uses the YunoHost postfix server for SMTP, and uses the administrator e-mail as sender. There's no need for authentication to send e-mails from the server.
So, that would need a separate dedicated mailbox so that the administrator doesn't get spammed by all these "reply-by-email" emails. Plus you need to provide credentials to read fetch e-mails.

But, you're damn right: Discourse doesn't seem to support IMAP access to read e-mails yet! And YunoHost doesn't support accessing e-mails via POP3 out of the box. So that sounds like you would need to activate POP3 access (I don't see a known YunoHost ticket for this), or use another e-mail account somewhere else...

**Comment by tomdereub** [comment]
I did this to configure pop3 : https://forum.yunohost.org/t/how-to-enable-pop3-in-yunohost/1662/2
I don't know if it's working, but for the moment I can't receive any mail from discourse. It was working before I tried the pop3 configuration, but now I always have this kind of messages in the logs : 
Jul 30 12:51:09 lamo postfix/smtpd[493]: NOQUEUE: reject: RCPT from localhost[::1]: 553 5.7.1 <reponse+verp-b...@d.plaf.fr>: Sender address rejected: not logged in; from=<reponse+verp-b...@d.plaf.fr> to=<...> proto=ESMTP helo=<d.plaf.fr>

**Comment by JimboJoe** [comment]
Well, I suppose the "+verp-p" is a suffix added by discourse?
As stated in the Discourse howto, the email server must support the [plus addressing aka address tags](https://en.wikipedia.org/wiki/Email_address#Address_tags), which is maybe not the case on YunoHost at the moment. From what I see, there are already settings that seem related to that ([here](https://github.com/YunoHost/yunohost/blob/b3b3062d62642dd6c00aa44484401e94ac9e6128/data/templates/postfix/main.cf#L65), quite relevant according to [this tutorial](https://www.stevejenkins.com/blog/2011/03/how-to-use-address-tagging-usertagexample-com-with-postfix/)...)

**Comment by tomdereub** [comment]
Résultats des essais du jour, et de la discussion avec JimboJoe sur le chat : 
- pour récapituler, j'ai activé pop3 sur yunohost, et créé un user dédié qui s'appelle "reponse", et qui a donc un compte mail en reponse@mondomaine.
- tant qu'on n'active pas "reply by email enabled" dans discourse, les mails sont envoyés avec le "notification email", et ça marche bien (on peut mettre ce qu'on veut...).
- Si on active "reply by email enabled", si le modèle contient un truc genre "reponse+%{reply_key}@mondomaine", les mails ne partent plus, ils sont bloqués par postfix avec une erreur du type : `NOQUEUE: reject: RCPT from localhost[::1]: 553 5.7.1 <reponse+verp-f95deabd99ed1566865b83a1228cfc56@mondomaine>: Sender address rejected: not logged in`
- Si le modèle contient juste "%{reply_key}@mondomaine", on reçoit bien les mails quand quelqu'un répond à un sujet suivi, mais si on fait juste "répondre" ça ne marche pas, car %{reply_key}@mondomaine ne correspond à aucun compte mail. Par contre si on rajoute "reponse+" à la main au début de l'adresse en répondant ça marche, et ça s'ajoute dans le sujet sur le forum.

Pistes de résolutions envisagées : 
- faire en sorte que le smtp laisse passer les mails avec un +machin dans l'expéditeur, comme ça marche pour la réception
- faire en sorte que discourse envoie toujours les mails avec le "notification email" et ne mette le "reponse+%{reply_key}@mondomaine" que dans le champ "réponse à ".

**Comment by tomdereub** [comment]
I don't understand this : 
> I then changed my site email smtp config to user to forum@mysite.com with the relevant pass.
Is it in discourse config or in postfix config ?

**Comment by JimboJoe** [comment]
It's in `/var/www/discourse/config/discourse.conf`, in the `smtp_user_name` and `smtp_password` field.

Currently we don't use any authentication, but maybe authenticating with the dedicated Discourse user will do the trick...?

**Comment by tomdereub** [comment]
Yes ! I tried quite a lot of different sets of parameters, and finally this works well : 
```
# address of smtp server used to send emails
smtp_address = mondomaine                               #it's my main yunohost domaine

# port of smtp server used to send emails
smtp_port = 25                                                   #unchanged

# domain passed to smtp server
smtp_domain = d.mondomaine                           #discourse domain

# username for smtp server
smtp_user_name = reponse                                #username of yunohost user I created for discourse (and wich is on d.mondomaine)

# password for smtp server
smtp_password = reponse's password

# smtp authentication mechanism
smtp_authentication = plain                               #unchanged

# enable TLS encryption for smtp connections
smtp_enable_start_tls = true                               #set to true
```
Thanks a lot ! Now it would be cool to make a "how to" on the forum, maybe I can take the time to do that. And much bettet, is it be possible to integrate this with the installation of discourse_ynh ? Create the user, and configure discourse configuration.

**Comment by JimboJoe** [comment]
Very good news, thanks a lot for your persistence!! :+1: 
For the automation, I think @Jibec already tried to create a dedicated user in a YunoHost package, but I think he experienced a lock in the `yunohost` command... To be confirmed or tried :thinking:

**Comment by Jibec** [comment]
related issue is: https://github.com/YunoHost/issues/issues/947

I tried to make it work and finally stoped. What we have to do is to edit ldap but I'm unsure how to do this without breaking the system. 

Here is what postfix search for (`/etc/postfix/ldap-accounts.cf`)
```
server_host = localhost
server_port = 389
search_base = dc=yunohost,dc=org
query_filter = (&(objectClass=mailAccount)(mail=%s))
result_attribute = uid
/etc/postfix/ldap-accounts.cf (END)
```

When you search into YunoHost's LDAP: `ldapsearch -LLL -D cn=admin,dc=yunohost,dc=org -W -b dc=yunohost,dc=org`, you'll get everything.

Some documentation here: https://github.com/YunoHost/moulinette/blob/06626f69c40de7a2ee25e6c1d6ff95077954301a/doc/ldap.rst#ldap-schema

You'll notice "maildrop" attributes. It is something you'll get filled if you have an email redirection.

What I think we should do for an app needed SMTP:
1. create an "$app" user in LDAP,
2. add a "maildrop" attribute with the real YunoHost user who should receive emails,
3. create "mail" attribut for every addresses the app need

Basically, we need to do this:
https://github.com/YunoHost/moulinette/blob/06626f69c40de7a2ee25e6c1d6ff95077954301a/doc/ldap.rst#adding-user-in-ldap
but without "inetOrgPerson"

If someone knows how to really build this, I would love to help.
Remember to play with it carefully, you can be sure to be able to misconfigure servers  by not handling LDAP well.

**Comment by JimboJoe** [comment]
Thanks @Jibec .
What you relate is the lack of DKIM for e-mail sent without authenticating with Postfix, which is indeed needed for most apps!

What we're considering in this issue is creating a dedicated e-mail, because Discourse needs to be able to actually read the e-mails to handle the reply-by-email feature (and we don't want to mix-up with the e-mails of a "real" person).
Thus it would be interesting to automate the creation of a YunoHost user (hence with a mailbox). I think you tried to create a YunoHost user for an app (was it Pagure or Weblate?), but it didn't work... Do you remember anything like that?
Thanks!

**Comment by tomdereub** [comment]
Just to inform, the app upgrade I made today has overriden discourse.conf. So I hade to change it again to make mail work.

**Comment by JimboJoe** [comment]
Thanks for the report.
There's currently no easy way to handle upgrades while keeping configurations (an update could bring new configuration entries that could prevent the app from working if not present).
You should nonetheless have been warned that your configuration file had been spotted as modified and backed up prior to being overwritten. Was it the case?

**Comment by decentral1se** [comment]
Hmmm, I am struggling with this. I've tried to follow along but I'm now seeing this error:

```
dovecot: lda(response@foo.org): Error: sieve: msgid=<...responsekey...@foo.org>: failed to store into mailbox '...responsekey...': Mailbox doesn't exist: ...responsekey...
```

I've created a yunohost user "response" and configured the reply by email to use it. When I reply by email, the emails do arrive in the inbox of the "response" user which I check with Rainloop. However, they do not make it to discourse ... 

I've also install and configured pop3 (although I now realise it is not necessary). So far, I can send outgoing emails well (and have configured Mailjet for that) but no replies by email.

**Comment by decentral1se** [comment]
OK, it's working! Big thanks to you all. 

Here's a documentation PR https://github.com/YunoHost-Apps/discourse_ynh/pull/32!

**Comment by decentral1se** [comment]
> So that sounds like you would need to activate POP3 access (I don't see a known YunoHost ticket for this),

Opened up https://github.com/YunoHost/issues/issues/1393 about this!

**Comment by nicolasVr** [comment]
I have been struggling for some days with Discourse on Yunohost smtp issues.
Since an update in the middle of July my Discourse doesn't send anymore any email. The config I had setup doesn't work anymore.
I think something changed in Yunohost since before it was possible to set a main adress foo2@domain.com for user foo1 while now it must be foo1@domain.com
Anyway I have tried any possible configs in discourse.conf for stmp and always have in the logs 
`Job exception: 535 5.7.8 Error: authentication failed: `
I am quite desperate as this is a big problem for the organisation I am involved in. I would like to know if some of you managed to have this working ? If yes, what is your config ?
